server/users.ts

```diff
@@ -1040,12 +1040,19 @@ export class User extends Chat.MessageContext {
 			Rooms.get(roomid)!.onLeave(oldUser);
 		}
 
-		if (this.locked === '#dnsbl' && !oldUser.locked) this.locked = null;
-		if (!this.locked && oldUser.locked === '#dnsbl') oldUser.locked = null;
+		const startLock = this.locked;
+		const startSemilock = this.semilocked;
+
+		if (this.semilocked === '#dnsbl' && !oldUser.semilocked) this.semilocked = null;
+		if (!this.semilocked && oldUser.semilocked === '#dnsbl') oldUser.semilocked = null;
+		if ((this.locked === '#hostfilter' || this.locked === '#ipban') && !oldUser.locked) this.locked = null;
```

Comment:
Maybe just check for the first character to be `#`.

Reply:
Actually, maybe we should just forgive all locks that aren't username locks? We've definitely had problems with people in schools being perpetually locked. Is there a way to tell if a lock comes from an IP or a username?

Reply:
I'm somewhat wary of special locks starting with ``#`` and whether they should be treated the same way.  I know ``#proxy`` and ``#range`` are other ones.

Reply:
> Actually, maybe we should just forgive all locks that aren't username locks? We've definitely had problems with people in schools being perpetually locked. Is there a way to tell if a lock comes from an IP or a username?

I think this would be too forgiving.  We already markshared schools, but if we forgave everyone who is locked on the same IP as an alt, we'd be letting a lot of actual alts go free.

Reply:
> I'm somewhat wary of special locks starting with ``#`` and whether they should be treated the same way.  I know ``#proxy`` and ``#range`` are other ones.  I think overall, it should be safe.

I can't think of one that shouldn't be freed, at least.

> I think this would be too forgiving. We already markshared schools, but if we forgave everyone who is locked on the same IP as an alt, we'd be letting a lot of actual alts go free.

Would we? The user would still need a new account _and_ a new IP, and if they have those, they could have evaded just by being more cautious anyway. I think a lot of schools aren't markshared, so I think it would be worth having a better experience for people on shared IPs, in exchange for slightly more evasion.

I don't have a good sense of how markshared works. Are schools locked with `#shared` or something?

Reply:
I get what you're talking about now.
> Is your idea that if an uninvolved name logs into a bad IP, it'll instantly forgive them if they log into a clean IP? While alts who were present during the lock are locked by name and remain locked?
I think this is fine since we already forgive random names that log into punished IPs; they simply have to log out and go onto another network, and we can remove the logout step.